### ISSUE TITLE
[CP] Cherry pick migrator to upgrade Gradle version when conflict with Android Studio bundled java is detected

### ISSUE THREAD

**Issue by gmackall** [flutter/flutter#125693]
### issue_link

https://github.com/flutter/flutter/issues/122376

### Commit Hash

eba2a52

### Target

stable

### pr_link

https://github.com/flutter/flutter/pull/126093

### Impacted Users

Flutter android devs who haven't upgraded the Gradle version in their project in ~1 year, and have upgraded to the latest stable Android Studio version ('Flamingo', or 2022.2.1).

### Impact Description

I would recommend looking at the issue that is linked from the PR (https://github.com/flutter/flutter/issues/122376) as it has a particularly good description of the impact, BUT with the important caveat that the original issue states that the error will be an inscrutable one of the form:

```
FAILURE: Build failed with an exception.

* Where:
Build file '…/example/android/build.gradle'

* What went wrong:
Could not compile build file '…/example/android/build.gradle'.
> startup failed:
  General error during conversion: Unsupported class file major version 61

  java.lang.IllegalArgumentException: Unsupported class file major version 61
  	at groovyjarjarasm.asm.ClassReader.<init>(ClassReader.java:189)
  	at groovyjarjarasm.asm.ClassReader.<init>(ClassReader.java:170)
  	[…
  	 …
  	 … 209 more lines of Groovy and Gradle stack trace …
  	 …
  	 …]
  	at java.base/java.lang.Thread.run(Thread.java:833)
```
and this is actually no longer the case thanks to the work in: https://github.com/flutter/flutter/pull/124084. Rather the error will inform them of the actual issue, and provide a link to a flutter documentation page with mitigation steps.

The summary is that `flutter run` and `flutter build` for android will error for flutter devs who 
1. upgrade to the Android Studio 'Flamingo' stable version released this month (or are otherwise using a newer version).
2. last ran `flutter create` over ~1 year ago, and haven't updated the gradle version in their gradle-wrapper.properties file since.

### Workaround

Yes, the workaround would be for the devs to follow the migration steps and directly upgrade the gradle version in their gradle-wrapper.properties file. 

### Risk

high

### Test Coverage

yes

### Validation Steps

To validate that this fix works, it would be best to
1. Ensure that your android studio version is at least the latest stable ('Flamingo', or 2022.2.1).
2. Using a version of flutter older than 3.0 (so 2.10.5 for example, commit hash 5464c5b), run `flutter create`.
3. For a quick sanity check: upgrade to a recent version of flutter that _does not_ contain this fix (commit hash fc2098 for example) and run `flutter run` with an android device or emulator. You should see an error with output towards the bottom of 
```
┌─ Flutter Fix ─────────────────────────────────────────────────────────────────────────────────────────────┐
│ [!] Your project's Gradle version is incompatible with the Java version that Flutter is using for Gradle. │
│                                                                                                           │
│ To fix this issue, consult the migration guide at docs.flutter.dev/go/android-java-gradle-error.          │
└───────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```
4. Switch to using a version of flutter containing this fix (e.g. master).
5. Run `flutter run` and look for a log message noting that your gradle version is being upgraded*.
6. Confirm by checking that the version of gradle in your gradle-wrapper.properties has been upgraded to 7.6.1 (from the root of your project, open ./android/gradle/wrapper/gradle-wrapper.properties and look at the line that starts with distributionUrl).

* Note on step 5 - for me `flutter run` actually still failed, but for a separate issue that would have needed fixing afterwards in this example project regardless (in this case, changing the AGP version in android/build.gradle to a newer one, for example 7.3.0). After fixing this `flutter run` succeeds.

**Comment by itsjustkevin** [comment]
Thanks for this cherry-pick @gmackall, however it seems a bit risky.  Reaching out to @reidbaker and @CaseyHillers for another look.

**Comment by CaseyHillers** [comment]
@gmackall in the PR link, you'll need to open a CP request PR to the candidate branch so we can review this request.

**Comment by reidbaker** [comment]
I think we should take it. Lets talk on monday.

**Comment by gmackall** [comment]
> @gmackall in the PR link, you'll need to open a CP request PR to the candidate branch so we can review this request.

I see, I misunderstood https://github.com/flutter/flutter/wiki/Flutter-Cherrypick-Process, and thought the pr_link was to the original PR. I'll create a cherry pick review and edit the link to fix this.

**Comment by reidbaker** [comment]
This pr should not be accepted without https://github.com/flutter/flutter/pull/125728.

**Comment by itsjustkevin** [comment]
@reidbaker and @gmackall seems like this has fallen to the wayside, possible hold for the first stable hotfix.

**Comment by reidbaker** [comment]
> @reidbaker and @gmackall seems like this has fallen to the wayside, possible hold for the first stable hotfix. 

I would very much disagree that this is fallen by the wayside. This has been @gmackall top priority for at least 2 weeks. The issue was that we found a bug which made us roll back. The original PR and a fixed PR is now up. 

That said, I can understand if we are too close to the release to cherry pick something in but my impression was that was Friday of this week.

**Comment by reidbaker** [comment]
That makes total sense and I trust yall to make the right decision I just wanted to make sure it was clear quickly that this is not something that has been paused for more urgent work.

**Comment by itsjustkevin** [comment]
Thanks @reidbaker!  Also apologies @gmackall, no ill intent here.

**Comment by reidbaker** [comment]
@gmackall please keep @itsjustkevin up to date on this progress of this CP every day from now until it is either accepted or rejected.

**Comment by gmackall** [comment]
> Thanks @reidbaker! Also apologies @gmackall, no ill intent here.

No worries! Sorry I should have been providing updates here alongside the updates to the original PRs. I'll make sure this is updated with status at all times from now on!

I am about to merge the [patched version](https://github.com/flutter/flutter/pull/125836#pullrequestreview-1409271525) of the original reverted PR, and then add the cherry pick PR to this issue.

**Comment by gmackall** [comment]
Still waiting on the autosubmit to land the patched version, then will make the cherry pick PR.

**Comment by gmackall** [comment]
Patched version is merged, going to monitor post submits for a little bit before creating the cherry pick PR

**Comment by gmackall** [comment]
I did not notice any issues with post submits, am working on the cherry pick PR at the moment.

It unfortunately has some issues in that the original PR somewhat depends on two un-cherry-picked changes,
1. A change to make [android studio version nullable](https://github.com/flutter/flutter/pull/124771) (instead of being set to unknown)
2. A change that (among other things) [pulls globals into gradle.dart](https://github.com/flutter/flutter/pull/123034)

This is my first cherry pick request, in situations like this is the preferred way to resolve to pull in the minimal set of changes that your PR depends on, or is there another best way (or does it typically disqualify a cherry pick if it depends on additional un-cherrypicked commits)?

Either way at the moment I am trying the former!

**Comment by CaseyHillers** [comment]
Please avoid pulling in "the minimal changes." There's less risk if we CP in the other 2 commits as well, which you can add as part of this CP request.

**Comment by gmackall** [comment]
Got it, that makes sense. Thanks!

**Comment by gmackall** [comment]
I've corrected the pr link in the issue to point to a cherry pick that includes the original migrator PR. It also ended up taking 3 additional commits (I've linked all of the original PRs in the cherry-pick PR).

For some reason the CLA check is failing, it looks like maybe the git cherry-pick lost some context on the contributors because the [failing check](https://github.com/flutter/flutter/pull/126093/checks?check_run_id=13248269331) has the same list of authors as the [original succeeding check](https://github.com/flutter/flutter/runs/12733623908) (from the non cherry-pick PR).

**Comment by gmackall** [comment]
The CLA issue is fixed and the cherry pick should be ready for review, though I understand that the risk is high now being this late (and that it requires 4 commits).

**Comment by CaseyHillers** [comment]
@reidbaker can you review this CP request?

**Comment by reidbaker** [comment]
> @reidbaker can you review this CP request?

Comment added on the PR. I think we should accept this but do it as a patch for 3.10 and not rush it today.

**Comment by CaseyHillers** [comment]
@reidbaker @gmackall are we comfortable with landing this in next week's stable hotfix, or do we need more soak time?

**Comment by reidbaker** [comment]
Defer to @gmackall. We have not had any issues I am aware of.

**Comment by gmackall** [comment]
> @reidbaker @gmackall are we comfortable with landing this in next week's stable hotfix, or do we need more soak time?

I am comfortable with that timeline. I am happy with the amount of soak time we have had and haven't seen any issues.

**Comment by XilaiZhang** [comment]
landed on stable
